### PR TITLE
[RLMObjectSchema] +dynamicSchemaForRealm: now respects search indexes

### PR DESCRIPTION
**Pull request by jpsim**
/cc @tgoyne @alazier

### CODE REVIEW

File: Realm/RLMProperty.mm [hunk 1]
```diff
@@ -306,6 +306,7 @@ - (id)copyWithZone:(NSZone *)zone {
 
 - (BOOL)isEqualToProperty:(RLMProperty *)property {
     return _type == property->_type
+	&& _indexed == property->_indexed
```
Comment (segiddins) [inline Realm/RLMProperty.mm]: is this indentation off?

File: Realm/RLMObjectSchema.mm [hunk 2]
```diff
@@ -285,11 +285,8 @@ - (BOOL)isEqualToObjectSchema:(RLMObjectSchema *)objectSchema {
     NSArray *otherProperties = objectSchema.properties;
     for (NSUInteger i = 0; i < _properties.count; i++) {
         RLMProperty *p1 = _properties[i], *p2 = otherProperties[i];
-        if (p1.type != p2.type ||
```
Comment (tgoyne) [inline Realm/RLMObjectSchema.mm]: We don't want to be checking for `indexed` being equal in the place this is used (checking if we can use the same accessor classes).

**Comment by tgoyne** [comment]
This needs a test that explicitly verifies that the `indexed` property is being read property.

File: Realm/RLMObjectSchema.mm (hunk 2 above)
Comment (jpsim) [inline Realm/RLMObjectSchema.mm]: But we do want to check for `isPrimary`? Can you explain why, I don't understand the distinction wrt accessors.

Comment (tgoyne) [inline Realm/RLMObjectSchema.mm]: The setter for primary keys checks for uniqueness.

**Comment by alazier** [comment]
I'm not sure why this pr is needed at all. What is the problem that you are trying to solve? I don't think indexed shouldn't be used to determine property or schema equality at all. Add/removing an index does not change the schema, and should not require migrations.

**Comment by jpsim** [comment]
> Add/removing an index does not change the schema, and should not require migrations.

Shouldn't adding an index to a property add an index to that column and therefore require a migration?

The main goal of this PR is to make `+[RLMSchema dynamicSchemaForRealm:]` actually include the indexing settings present in the file in the schema. The other change is really just to simplify testing. I could leave the check for `indexed` out of `isEqualToProperty:` but it really belongs in there IMO.

**Comment by tgoyne** [comment]
Adding a property to `indexedProperties` should make it add an index to existing Realm files (and I think it currently doesn't), but adding or removing a search index doesn't require any migration logic, so like adding new tables it should be allowed without a schema version bump.

**Comment by alazier** [comment]
Another issue is that you can't tell if there is an index on property from a dynamic schema due to the property being primary, or if the user explicitly marked the property as indexed. The real solution to this problem is to actually have an indexedProperties metadata table to store this information.

**Comment by tgoyne** [comment]
When does it matter that the user redundantly specified that the pk should be indexed?

**Comment by jpsim** [comment]
Regardless of *why* the property has an index, its `indexed` property should be set.

**Comment by alazier** [comment]
@tgoyne it doesn't matter until we make the dynamic api public. At that point if a user change the primary key we need to know whether the property should still be indexed or not. Right now I don't think there is not a scenario where this can occur. 

I'm fine with the change of populating the indexed property for dynamic schema so :+1:

**Comment by tgoyne** [comment]
Even with the dynamic API I don't really see how it'd be an issue. If they give us an `RLMProperty` with `primary = NO` and `indexed = YES` then it'll stay indexed when we make it not a pk, and if it has `indexed = NO` the index should be removed. I don't think it makes sense to store some metadata that tries to cleverly override what the user tells us to do based on what we think they intended.

**Comment by tgoyne** [comment]
The updated test still doesn't actually verify that the indexed property is read correctly or that `isEqualToProperty:` is correct; it just verifies that either neither or both of them are incorrect.

**Comment by alazier** [comment]
I don't think it makes sense to worry about any of this right now as there are clearly higher priorities at the moment.

**Comment by jpsim** [comment]
> I don't think it makes sense to worry about any of this right now as there are clearly higher priorities at the moment.

I really just needed this because I had to use the dynamic API to help debug a support issue. I'll address @tgoyne's point about adding a test for `isEqualToProperty:` and then hopefully we can get this in.

**Comment by alazier** [comment]
That's fair enough. I'm trying to get all of the other swift stuff rebased/merged so don't have the time to debate this now. I'm fine with the current pr - isEqualToProperty isn't used internally so I'm not too worried about it, but @tgoyne is probably correct that it should be tested - we could also just abandon that change as well for the time being if it a lot of effort.

**Comment by jpsim** [comment]
`isEqualToProperty:` now has some tests. I could spend more time covering more cases, but that doesn't seem worthwhile. I could keep this PR open and add more tests later if you prefer. @tgoyne @alazier

**Pull request merged**